### ISSUE TITLE
Retrieve UTF-8 JSONObject from Server 

### ISSUE THREAD

**Issue by FarshadKazemi** [koush/AndroidAsync#322]
Dear Koush

Almost I'm newbie in android programming ... 
I'm going to retrieve my JSONObject that is included Persian characters.
I use below code to retrieve my JSONObject:

    AsyncHttpPost post = new  AsyncHttpPost ("AAA");
    MultipartFormDataBody body = new MultipartFormDataBody();
    body.addStringPart("xxx","yyy");
    post.setBody(body);
    AsyncHttpClient.getDefaultInstance().executeJSONObject(post, new JSONObjectCallBack(){
    @override
    public void onCompleted(Exception e, AsyncHttpResponse response, JSONObject json)
    {
    Log.i("JSON Object:" , json.toString());
    }
    });

But Persian words aren't shown correctly ...
What should I do ?!

**Comment by koush** [comment]
Is it your server? Make sure you have Content-Type charset utf8

**Comment by FarshadKazemi** [comment]
Yeah, I have a PHP file that return data from mySQL as JSON object.
In my android application, when I use my JSONParser and ASYNCTASK classes, I get current JSONObject.

This is my JSONParser Class:
       
    public class JSONParser {
	static InputStream is = null;
	static JSONObject jObj = null;
	static String json = "";

	public JSONParser() {
	}

	/**
	 * Send POST Data to Server & Retrieve JSON From URL
	 */
	public JSONObject getJSONFromUrl(String url, List<NameValuePair> params) {
		// Making HTTP request
		try {
			// defaultHttpClient
			DefaultHttpClient httpClient = new DefaultHttpClient();
			HttpPost httpPost = new HttpPost(url);
			httpPost.setEntity(new UrlEncodedFormEntity(params,"UTF-8"));
			HttpResponse httpResponse = httpClient.execute(httpPost);
			HttpEntity httpEntity = httpResponse.getEntity();
			is = httpEntity.getContent();
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		} catch (ClientProtocolException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
		try {
			BufferedReader reader = new BufferedReader(new InputStreamReader(
					is, "UTF8"), 8);
			StringBuilder sb = new StringBuilder();
			String line = null;
			while ((line = reader.readLine()) != null) {
				sb.append(line + "\n");
			}
			is.close();
			json = sb.toString();
		} catch (Exception e) {
			Log.e("Buffer Error", "Error converting result " + e.toString());
		}
		try {
			jObj = new JSONObject(json);
		} catch (JSONException e) {
			Log.e("JSON Parser", "Error parsing data " + e.toString());
		}
		// return JSON Object
		return jObj;
	}

    }

**Comment by adzeitor** [comment]
http://www.ietf.org/rfc/rfc4627.txt

```
3.  Encoding

   JSON text SHALL be encoded in Unicode.  The default encoding is
   UTF-8.

   Since the first two characters of a JSON text will always be ASCII
   characters [RFC0020], it is possible to determine whether an octet
   stream is UTF-8, UTF-16 (BE or LE), or UTF-32 (BE or LE) by looking
   at the pattern of nulls in the first four octets.

           00 00 00 xx  UTF-32BE
           00 xx 00 xx  UTF-16BE
           xx 00 00 00  UTF-32LE
           xx 00 xx 00  UTF-16LE
           xx xx xx xx  UTF-8
```

Sorry for no time for patch ;)

**Comment by FarshadKazemi** [comment]
Dear Alexander 

Tanks so much for your hint

**Comment by qmtco** [comment]
tanks.
خییییلییییی ممنون
